### PR TITLE
graphql: add support for aliases

### LINKED ISSUE #5: graphql: add support for aliases for fields
The GraphQL spec supports **aliases** and the current parser that we have also recognises them.

```graphql

query {
     user: user_table {
          user_id: id
          name: user_name
     }
}

```

Either we have to 
- remap the JSON upon obtaining the result from the DB [which I think is slightly harder]

OR

- update the SQL gen function to accommodate queries with aliases (the magical `AS` keyword)

### PR DESCRIPTION
**Pull request by kolharsam**
this PR should resolve #5 

there's some clean up done on the PR to ensure that we process only 1 query root ONLY, since this is what happens on the hasura too.

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -25,44 +25,9 @@ pub fn string_to_static_str(s: String) -> &'static str {
...
-                    None => {},
-                }
-            }
-            new_json_arr.push(serde_json::Value::Object(new_obj));
-        }
-
-        return serde_json::Value::Array(new_json_arr);
+pub fn optional_fmap<T, U>(op: Option<T>, map_fn: fn(&T) -> U) -> Option<U> {
+    if let Some(t) = op {
+        return Some(map_fn(&t));
```
Comment (iykekings) [inline src/utils.rs:30]: There is a map fn on Options that we can use for this https://doc.rust-lang.org/std/option/enum.Option.html#method.map
So we can do 
```rust
optionalType.map(fn)
```

File: src/server.rs [hunk 2]
```diff
@@ -77,49 +78,78 @@ pub struct GraphQLRequest {
...
-    let mut query_fields: Vec<String> = Vec::new();
+    let mut fields_map: indexmap::IndexMap<FieldName, Vec<FieldName>> = indexmap::IndexMap::new();
 
     for set in qry.selection_set.items.iter() {
         if let graphql_parser::query::Selection::Field(field) = set {
             // FIXME: make this recursive too, this is just one level now...
             let table_name = field.name.to_string();
-            query_fields.append(&mut selection_set_fields_parser(&field.selection_set));
-            fields_map.insert(table_name, query_fields.clone());
+            let alias: Option<String> = utils::optional_fmap(field.alias, |&s| s.to_string());
```
Comment (iykekings) [inline src/server.rs:93]:
```suggestion
            let alias = field.alias.map(String::from);
```

File: src/server.rs [hunk 3]
```diff
@@ -77,49 +78,78 @@ pub struct GraphQLRequest {
...
-) -> Vec<String> {
-    let mut fields: Vec<String> = Vec::new();
+) -> Vec<FieldName> {
+    let mut fields: Vec<FieldName> = Vec::new();
 
-    // TODO: aliases are also not supported
     for set_item in sel_set.items.iter() {
         if let graphql_parser::query::Selection::Field(fld) = set_item {
-            fields.push(fld.name.to_string());
+            let alias: Option<String> = utils::optional_fmap(fld.alias, |&s| s.to_string());
```
Comment (iykekings) [inline src/server.rs:151]:
```suggestion
            let alias = fld.alias.map(String::from);
```

File: src/server.rs [hunk 4]
```diff
@@ -130,24 +160,27 @@ fn fetch_result_from_selection_set<'a>(
     sel_set: &graphql_parser::query::SelectionSet<'a, &'a str>,
     _client: &mut Client,
 ) -> actix_web::HttpResponse {
-    let mut fields_map: indexmap::IndexMap<String, Vec<String>> = indexmap::IndexMap::new();
+    let mut fields_map: indexmap::IndexMap<FieldName, Vec<FieldName>> = indexmap::IndexMap::new();
 
     for set_item in sel_set.items.iter() {
         if let graphql_parser::query::Selection::Field(fld) = set_item {
+            let alias: Option<String> = utils::optional_fmap(fld.alias, |&s| s.to_string());
```
Comment (iykekings) [inline src/server.rs:167]:
```suggestion
           let alias = fld.alias.map(String::from);
```

**Review by iykekings — CHANGES_REQUESTED** [review]
@kolharsam I left some comments on mapFn on Option. Nice work  🎉

File: src/utils.rs (hunk 1 above)
Comment (kolharsam) [reply src/utils.rs:30]: This is awesome! Thanks for letting me know about this! :)

**Comment by kolharsam** [comment]
@iykekings let me know if you have any other comments. :)
Please approve so that I can merge it :tada:

**Review by iykekings — APPROVED** [review]
Great work buddy! Sorry for the delay!

**Pull request merged**